File: solutions/r/rotational-cipher/rotational-cipher.R
# Given a plaintext and amount to shift by, return a rotated string.
rotate <- function(text, key) {
    do_shift <- function(c, n, values) {
        offset <- utf8ToInt(c) + n - utf8ToInt(values[1])
        values[1 + offset %% length(values)]
    }

    shift <- Vectorize(function(char) {
        if (char %in% letters) {
            do_shift(char, key, letters)
        } else if (char %in% LETTERS) {
            do_shift(char, key, LETTERS)
        } else {
            char
        }
    })

    (text |> strsplit("") |> unlist() |> shift() |> paste(collapse = ""))
}

File: solutions/elixir/rotational-cipher/lib/rotational_cipher.ex
defmodule RotationalCipher do
  @moduledoc false
  import Enum, only: [at: 2, map: 2]

  @doc """
  Given a plaintext and amount to shift by, return a rotated string.

  Example:
  iex> RotationalCipher.rotate("Attack at dawn", 13)
  "Nggnpx ng qnja"
  """
  @spec rotate(text :: String.t(), shift :: integer) :: String.t()
  def rotate(text, shift),
    do: text |> String.to_charlist() |> map(&rotate_char(&1, shift)) |> to_string()

  def rotate_char(c, n) when c in ?a..?z, do: ?a..?z |> at(rem(c + n - ?a, 26))
  def rotate_char(c, n) when c in ?A..?Z, do: ?A..?Z |> at(rem(c + n - ?A, 26))
  def rotate_char(c, _), do: c
end
